diff --git a/core/modules/contextual/contextual.module b/core/modules/contextual/contextual.module
index aaab3a6e9d..6f00ebfdb5 100644
--- a/core/modules/contextual/contextual.module
+++ b/core/modules/contextual/contextual.module
@@ -177,14 +177,19 @@ function _contextual_links_to_id($contextual_links) {
   $ids = [];
   $langcode = \Drupal::languageManager()->getCurrentLanguage(LanguageInterface::TYPE_URL)->getId();
   foreach ($contextual_links as $group => $args) {
-    $route_parameters = UrlHelper::buildQuery($args['route_parameters']);
-    $args += ['metadata' => []];
-    // Add the current URL language to metadata so a different ID will be
-    // computed when URLs vary by language. This allows to store different
-    // language-aware contextual links on the client side.
-    $args['metadata'] += ['langcode' => $langcode];
-    $metadata = UrlHelper::buildQuery($args['metadata']);
-    $ids[] = "{$group}:{$route_parameters}:{$metadata}";
+    try {
+      $route_parameters = ((isset($args['route_parameters']) && is_array($args['route_parameters'])) ? UrlHelper::buildQuery($args['route_parameters']) : UrlHelper::buildQuery([]));
+      $args += ['metadata' => []];
+      // Add the current URL language to metadata so a different ID will be
+      // computed when URLs vary by language. This allows to store different
+      // language-aware contextual links on the client side.
+      $args['metadata'] += ['langcode' => $langcode];
+      $metadata = UrlHelper::buildQuery($args['metadata']);
+      $ids[] = "{$group}:{$route_parameters}:{$metadata}";
+    }
+    catch (ClientException $e) {
+      \Drupal::logger('contextual')->warning('Contextual link for group @group has broken metadata for language @langcode code', ['@group' => $group, '@langcode' => $langcode]);
+    }
   }
   return implode('|', $ids);
 }
