diff --git a/core/modules/big_pipe/js/big_pipe.js b/core/modules/big_pipe/js/big_pipe.js
index 5d592e72f1..2229774612 100644
--- a/core/modules/big_pipe/js/big_pipe.js
+++ b/core/modules/big_pipe/js/big_pipe.js
@@ -89,10 +89,12 @@
    */
   function checkMutation(node) {
     return Boolean(
-      node.nodeType === Node.ELEMENT_NODE &&
+      node &&
+        node.nodeType === Node.ELEMENT_NODE &&
         node.nodeName === 'SCRIPT' &&
         node.dataset &&
         node.dataset.bigPipeReplacementForPlaceholderWithId &&
+        drupalSettings.bigPipePlaceholderIds &&
         typeof drupalSettings.bigPipePlaceholderIds[
           node.dataset.bigPipeReplacementForPlaceholderWithId
         ] !== 'undefined',
