diff --git a/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php b/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php
index 1d890c974d..f3001d5ed7 100644
--- a/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php
+++ b/core/modules/media_library/src/Plugin/Field/FieldWidget/MediaLibraryWidget.php
@@ -959,6 +959,10 @@ public static function validateRequired(array $element, FormStateInterface $form
     if (in_array([static::class, 'removeItem'], $form_state->getSubmitHandlers(), TRUE)) {
       return;
     }
+    // If the field is not accessible, this validation isn't needed.
+    if (isset($element['#access']) && !$element['#access']) {
+      return;
+    }
 
     $field_state = static::getFieldState($element, $form_state);
     // Trigger error if the field is required and no media is present. Although
